pip_audit/_service/esms.py

```diff
@@ -0,0 +1,165 @@
+            # Get the IDs, prioritising PYSEC and CVE.
+            ids: list[VulnerabilityID] = vuln["identifiers"]
+            ids.sort(key=type(self)._id_comparison_key)
+
+            # If the vulnerability has been withdrawn, we skip it entirely.
+            withdrawn_at = vuln["withdrawn_at"]
+            if withdrawn_at is not None:
+                logger.debug(f"ESMS vuln entry '{ids[0]}' marked as withdrawn at {withdrawn_at}")
+                continue
+
+            # The title is intended to be shorter, so we prefer it over
+            # description, if present. The Ecosyste.ms advisory metadata states that
+            # these fields *should* always be of type `str`; we are being defensive
+            # here and checking if the strings are empty.
+            description = vuln["title"]
+            if not description:
+                description = vuln["description"]
+            if not description:
+                description = "N/A"
+
+            # The "title" field should be a single line, but "description" might
+            # be multiple (Markdown-formatted) lines. So, we normalize our
+            # description into a single line (and potentially break the Markdown
+            # formatting in the process).
+            description = description.replace("\n", " ")
+
+            seen_vulnerable = False
+            fix_versions: set[Version] = set()
+            for affected in vuln["packages"]:
+                # We only care about PyPI versions.
+                if (
+                    affected["package_name"] != spec.canonical_name
+                    or affected["ecosystem"] != "pypi"
+                ):
+                    continue
+
+                for record in affected["versions"]:
+                    # Very silly: OSV version specs use single `=` for exact matches, while PEP 440
+                    # requires double `==`. All OSV operators have equivalent semantics to their
+                    # PEP 440 counterparts, so we do some gross regex munging here to accommodate for
+                    # the syntactical difference.
+                    osv_spec: str = record["vulnerable_version_range"]
+                    vulnerable = SpecifierSet(re.sub(r"(^|(, ))=", r"\1==", osv_spec))
+                    if not vulnerable.contains(spec.version):
+                        continue
+
+                    seen_vulnerable = True
+                    if (patched := record.get("first_patched_version")) is not None:
+                        fix_versions.add(Version(patched))
+                    break
+
+            if not seen_vulnerable:
+                continue
+
+            results.append(
+                VulnerabilityResult(
+                    id=ids[0],
+                    description=description,
+                    fix_versions=sorted(fix_versions),
+                    aliases=set(ids[1:]),
+                    published=self._parse_rfc3339(vuln.get("published")),
```

Comment:
Follow-up: let's create a `VulnerabilityResult.create(...)` or similar helper API that takes `ids, description, fix_versions, published` and splits `ids` into `id` and `aliases` as appropriate internally.